File: translater.js
const fs = require('fs');
const { exec } = require('child_process');
const { promisify } = require('util');

const execAsync = promisify(exec);

// Проверка аргументов командной строки
if (process.argv.length < 3) {
    console.error('Ошибка: Не указан файл для перевода');
    console.log('Использование: node script.js filename.srt [from:to]');
    console.log('Пример: node script.js movie.srt en:fr - перевод с английского на французский');
    console.log('По умолчанию: en:ru - перевод с английского на русский');
    process.exit(1);
}

const srtFilePath = process.argv[2];
let translationDirection = 'en:ru'; // Направление перевода по умолчанию

// Проверка наличия аргумента с направлением перевода
if (process.argv.length >= 4) {
    const directionArg = process.argv[3];
    if (/^[a-z]{2}:[a-z]{2}$/.test(directionArg)) {
        translationDirection = directionArg;
    } else {
        console.error('Ошибка: Неверный формат направления перевода. Используйте формат "from:to", например "en:fr"');
        process.exit(1);
    }
}

const [fromLang, toLang] = translationDirection.split(':');

// Проверка существования файла
if (!fs.existsSync(srtFilePath)) {
    console.error(`Ошибка: Файл ${srtFilePath} не найден`);
    process.exit(1);
}

console.log(`Файл \x1b[1;33m${srtFilePath}\x1b[0m`);
console.log(`Направление перевода: \x1b[1;36m${fromLang} → ${toLang}\x1b[0m`);
console.log('Начинаю перевод...');

async function translateFile() {
    const startTime = Date.now(); // Засекаем время начала

    try {
        // Чтение исходного файла
        const original = fs.readFileSync(srtFilePath, 'utf8');

        // Создание массива строк
        const arrayOriginStrings = original.split(/\r?\n/);

        // Создание массива с нумерацией строк
        const numberedLines = arrayOriginStrings.map((line, index) => [index, line]);

        // Определение строк для перевода
        const arrayForTranslate = numberedLines.filter(([index, line]) => {
            const trimmedLine = line.trim();
            return trimmedLine &&
                !/^\d+$/.test(trimmedLine) &&
                !/^\d{2}:\d{2}:\d{2},\d{3} --> \d{2}:\d{2}:\d{2},\d{3}$/.test(trimmedLine);
        });

        // Асинхронная функция для перевода текста
        async function translateText(text) {
            try {
                // Экранирование кавычек для командной строки
                const escapedText = text.replace(/"/g, '\\"');
                const command = `trans -brief -no-warn -from ${fromLang} -to ${toLang} "${escapedText}"`;

                const { stdout, stderr } = await execAsync(command);
                if (stderr) {
                    console.error(`Ошибка перевода: ${stderr}`);
                    return text;
                }
                return stdout.trim();
            } catch (error) {
                console.error(`Ошибка перевода текста: ${text}`, error.message);
                return text;
            }
        }

        // Асинхронный перевод всех строк
        const translationPromises = arrayForTranslate.map(async ([index, line]) => {
            const translated = await translateText(line);
            return { index, translated };
        });

        // Ожидание завершения всех переводов
        const translationResults = await Promise.all(translationPromises);

        // Обновление строк в numberedLines
        translationResults.forEach(({ index, translated }) => {
            numberedLines[index][1] = translated;
        });

        // Создание переведенного текста
        const arrayTransated = numberedLines.map(([index, line]) => line);
        const translatedText = arrayTransated.join('\n');

        // Переименование оригинального файла
        const newFilePath = srtFilePath + '_';
        fs.renameSync(srtFilePath, newFilePath);

        // Сохранение переведенного файла
        fs.writeFileSync(srtFilePath, translatedText, 'utf8');

        // Расчет времени выполнения
        const endTime = Date.now();
        const durationInSeconds = Math.round((endTime - startTime) / 1000);

        // Форматирование времени
        let durationString;
        if (durationInSeconds < 60) {
            durationString = `${durationInSeconds} сек`;
        } else {
            const minutes = Math.floor(durationInSeconds / 60);
            const seconds = durationInSeconds % 60;
            durationString = `${minutes} мин ${seconds} сек`;
        }

        // Обновление статуса в консоли
        process.stdout.write('\x1B[2A\x1B[2K'); // Перемещаем курсор на 2 строки вверх и очищаем
        console.log(`\x1b[1;32mПереведен за ${durationString}\x1b[0m`);

    } catch (error) {
        // Обновление статуса в консоли
        process.stdout.write('\x1B[2A\x1B[2K'); // Перемещаем курсор на 2 строки вверх и очищаем
        console.log(`\x1b[1;31mНе удалось перевести\x1b[0m`);
        console.error('Ошибка:', error.message);
        process.exit(1);
    }
}

// Запуск асинхронного процесса перевода
translateFile();

File: translater.py
import sys
import os
import re
import subprocess
import time
from typing import List, Tuple

# ANSI-коды для цветного вывода в консоль
class Colors:
    YELLOW = "\033[1;33m"
    CYAN = "\033[1;36m"
    GREEN = "\033[1;32m"
    RED = "\033[1;31m"
    WHITE = "\033[1;37m"
    RESET = "\033[0m"

# Проверка аргументов командной строки
if len(sys.argv) < 2:
    print(f"{Colors.RED}Ошибка: Не указан входной файл{Colors.RESET}")
    print("Использование: python translate_srt.py filename.srt [from:to]")
    print("Пример: python translate_srt.py movie.srt en:fr - перевод с английского на французский")
    print("По умолчанию: en:ru - перевод с английского на русский")
    sys.exit(1)

srt_file_path = sys.argv[1]
translation_direction = "en:ru"  # Направление перевода по умолчанию

# Проверка аргумента направления перевода
if len(sys.argv) >= 3:
    direction_arg = sys.argv[2]
    if not re.match(r"^[a-z]{2}:[a-z]{2}$", direction_arg):
        print(f"{Colors.RED}Ошибка: Неверный формат направления перевода. Используйте формат 'from:to', например 'en:fr'{Colors.RESET}")
        sys.exit(1)
    translation_direction = direction_arg

from_lang, to_lang = translation_direction.split(":")

# Проверка существования файла
if not os.path.exists(srt_file_path):
    print(f"{Colors.RED}Ошибка: Файл {srt_file_path} не найден{Colors.RESET}")
    sys.exit(1)

print(f"Файл: {Colors.YELLOW}{srt_file_path}{Colors.RESET}")
print(f"Направление перевода: {Colors.CYAN}{from_lang} → {to_lang}{Colors.RESET}")
print("Начинаю перевод...")
# Вывод начальной белой полосы прогресса (100 символов)
print(f"{Colors.WHITE}{'█' * 100}{Colors.RESET}")

def translate_text(text: str) -> str:
    """Перевод текста с использованием утилиты 'trans'."""
    try:
        # Экранирование кавычек для безопасного выполнения команды
        escaped_text = text.replace('"', '\\"')
        command = f'trans -brief -no-warn -from {from_lang} -to {to_lang} "{escaped_text}"'
        
        # Выполнение команды и захват вывода
        result = subprocess.run(
            command, shell=True, capture_output=True, text=True, check=True
        )
        return result.stdout.strip()
    except subprocess.CalledProcessError as e:
        print(f"{Colors.RED}Ошибка перевода: {e.stderr}{Colors.RESET}")
        return text
    except Exception as e:
        print(f"{Colors.RED}Ошибка перевода текста '{text}': {str(e)}{Colors.RESET}")
        return text

def translate_file():
    """Основная функция для обработки и перевода файла .srt."""
    start_time = time.time()  # Засекаем время начала

    try:
        # Чтение исходного файла
        with open(srt_file_path, "r", encoding="utf-8") as f:
            original = f.read()

        # Разделение на строки
        lines = original.splitlines()

        # Создание списка кортежей (индекс, строка)
        numbered_lines: List[Tuple[int, str]] = [(i, line) for i, line in enumerate(lines)]

        # Фильтрация строк, которые нужно перевести
        lines_to_translate = [
            (index, line)
            for index, line in numbered_lines
            if line.strip()
            and not re.match(r"^\d+$", line.strip())
            and not re.match(r"^\d{2}:\d{2}:\d{2},\d{3} --> \d{2}:\d{2}:\d{2},\d{3}$", line.strip())
        ]

        total_lines = len(lines_to_translate)
        translated_lines = []

        # Перевод каждой строки с обновлением прогресс-бара
        for i, (index, line) in enumerate(lines_to_translate, 1):
            translated = translate_text(line)
            translated_lines.append((index, translated))
            
            # Вычисление прогресса
            progress = i / total_lines
            green_blocks = int(progress * 100)
            white_blocks = 100 - green_blocks
            
            # Обновление прогресс-бара (перемещение в начало строки)
            print(f"\033[1A\033[2K", end="")  # Подняться на 1 строку и очистить
            print(f"{Colors.GREEN}{'█' * green_blocks}{Colors.WHITE}{'█' * white_blocks}{Colors.RESET}")

        # Обновление исходных строк переведенным текстом
        for index, translated in translated_lines:
            numbered_lines[index] = (index, translated)

        # Создание переведенного текста
        translated_text = "\n".join(line for _, line in numbered_lines)

        # Переименование исходного файла
        backup_path = srt_file_path + "_"
        os.rename(srt_file_path, backup_path)

        # Сохранение переведенного файла
        with open(srt_file_path, "w", encoding="utf-8") as f:
            f.write(translated_text)

        # Расчет времени выполнения
        duration_seconds = round(time.time() - start_time)
        if duration_seconds < 60:
            duration_str = f"{duration_seconds} сек"
        else:
            minutes = duration_seconds // 60
            seconds = duration_seconds % 60
            duration_str = f"{minutes} мин {seconds} сек"

        # Замена прогресс-бара сообщением об успешном завершении
        print(f"\033[1A\033[2K", end="")  # Подняться на 1 строку и очистить
        print(f"{Colors.GREEN}Переведено за {duration_str}{Colors.RESET}")

    except Exception as e:
        # Обновление статуса в консоли
        print(f"\033[1A\033[2K", end="")  # Подняться на 1 строку и очистить
        print(f"{Colors.RED}Не удалось перевести{Colors.RESET}")
        print(f"Ошибка: {str(e)}")
        sys.exit(1)

if __name__ == "__main__":
    translate_file()


"""
Описание скрипта:

Этот скрипт предназначен для перевода текста в файлах субтитров формата .srt с одного языка на другой с использованием утилиты командной строки `trans`. Скрипт сохраняет структуру субтитров, переводя только текстовые строки, оставляя нетронутыми номера субтитров и временные метки. Исходный файл сохраняется с суффиксом `_1` (например, `movie.srt_1`), а переведенный текст записывается в файл с исходным именем.

### Основные функции:
1. **Проверка входных данных**: Скрипт принимает путь к файлу .srt и, опционально, направление перевода в формате `from:to` (например, `en:fr`). По умолчанию используется перевод с английского на русский (`en:ru`).
2. **Фильтрация строк**: Переводятся только текстовые строки, исключая пустые строки, номера субтитров и временные метки.
3. **Перевод**: Используется утилита `trans` для перевода каждой строки. Ошибки перевода логируются, и в случае сбоя сохраняется исходный текст.
4. **Прогресс-бар**: Во время перевода отображается шкала из 100 символов, которая начинается белой и постепенно становится зеленой по мере обработки строк.
5. **Сохранение результата**: После завершения перевода исходный файл переименовывается, а переведенный текст сохраняется. Прогресс-бар заменяется сообщением о времени выполнения (например, "Переведено за 3 сек").
6. **Обработка ошибок**: При возникновении ошибок (например, файл не найден или сбой перевода) выводится сообщение об ошибке, и процесс завершается.

### Использование:
1. **Убедитесь, что утилита `trans` установлена**:
   Установите `translate-shell` с помощью команды:
   ```bash
   sudo apt install translate-shell  # Для Ubuntu/Debian
"""
